### PR TITLE
Add an examples folder for code downstream tasks

### PR DESCRIPTION
**Pull request by loubnabnl**
# What does this PR do?
This PR adds a folder in CodeParrot directory to store examples for downstream tasks on code models.

### CODE REVIEW

File: examples/research_projects/codeparrot/examples/train_complexity_predictor.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
+
+    # Prepare the optimizer and learning rate scheduler
+    optimizer = AdamW(get_grouped_params(model, args), lr=args.learning_rate)
+    lr_scheduler = get_scheduler(
+        name=args.lr_scheduler_type,
+        optimizer=optimizer,
+        num_training_steps=args.num_epochs,
+        num_warmup_steps=args.num_warmup_steps,
+    )
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/examples/train_complexity_predictor.py:120]: I think you can define all that in the training arguments, no? No need to pass an optimizer/lr_scheduler explicitely?

File: examples/research_projects/codeparrot/examples/train_complexity_predictor.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    params_with_wd, params_without_wd = [], []
+    for n, p in model.named_parameters():
+        if any(nd in n for nd in no_decay):
+            params_without_wd.append(p)
+        else:
+            params_with_wd.append(p)
+    return [
+        {"params": params_with_wd, "weight_decay": args.weight_decay},
+        {"params": params_without_wd, "weight_decay": 0.0},
+    ]
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/examples/train_complexity_predictor.py:55]: I think if you don't pass the optimizer explicitly, the Trainer will take care of that for you, no?

File: examples/research_projects/codeparrot/examples/train_complexity_predictor.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+class CustomCallback(TrainerCallback):
+    def __init__(self, trainer) -> None:
+        super().__init__()
+        self._trainer = trainer
+
+    def on_epoch_end(self, args, state, control, **kwargs):
+        if control.should_evaluate:
+            control_copy = deepcopy(control)
+            self._trainer.evaluate(eval_dataset=self._trainer.train_dataset, metric_key_prefix="train")
+            return control_copy
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/examples/train_complexity_predictor.py:67]: Why is this needed? Isn't this the same as `evaluation_strategy="epoch"` in the training arguments? also why do you evaluate on the train set?

File: examples/research_projects/codeparrot/README.md [hunk 4]
```diff
@@ -12,7 +12,7 @@ This is an open-source effort to train and evaluate code generation models. Code
 - continuously push checkpoints to the hub with `huggingface_hub`
 - stream the dataset with `datasets` during training to avoid disk bottlenecks
 - apply the `code_eval` metric in `datasets` to evaluate on [OpenAI's _HumanEval_ benchmark](https://huggingface.co/datasets/openai_humaneval)
-
+- showcase examples for downstream tasks with code models in [examples](https://github.com/huggingface/transformers/tree/main/examples/research_projects/codeparrot/examples) folder
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/README.md:15]: maybe we can say what examples we show there. should we also add the code for the text2py and py2text here?

**Review by lvwerra — COMMENTED** [review]
Looks good, I mostly left comments for simplifying the training code a bit.

File: examples/research_projects/codeparrot/examples/train_complexity_predictor.py (hunk 1 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/examples/train_complexity_predictor.py:120]: I think by default the linear scheduler is used and I needed cosine scheduler here, but you're right we don't need to specify the optimizer

File: examples/research_projects/codeparrot/examples/train_complexity_predictor.py (hunk 3 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/examples/train_complexity_predictor.py:67]: I added this because I wanted to monitor the gap in accuracy between the training set and evaluation set

File: examples/research_projects/codeparrot/README.md (hunk 4 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/README.md:15]: Ok but the text2py and py2text examples use a similar script to the one for pretraining codeparrot just with a different dataset and model checkpoint, maybe I can just mention that in the README?

**Review by lvwerra — APPROVED** [review]
